File: built-in plugins dsp/mb-master-chain.cpp
/**
 * MB Master Chain
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : All-in-one mastering chain with EQ, compression, and limiting
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MASTER_CHAIN_H
#define MB_MASTER_CHAIN_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMasterChain : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-master-chain";
    static constexpr const char* PLUGIN_NAME    = "MB Master Chain";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float eqLow = 0f;  // range [-6, 6]
    float eqHigh = 0f;  // range [-6, 6]
    float compThreshold = -12f;  // range [-30, 0]
    float limCeiling = -0.3f;  // range [-3, 0]
    float output = 0f;  // range [-12, 12]
    };

    MbMasterChain() = default;
    ~MbMasterChain() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.eqLow = std::clamp(params.eqLow, -6f, 6f);
        params.eqHigh = std::clamp(params.eqHigh, -6f, 6f);
        params.compThreshold = std::clamp(params.compThreshold, -30f, 0f);
        params.limCeiling = std::clamp(params.limCeiling, -3f, 0f);
        params.output = std::clamp(params.output, -12f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Master Chain
        return input;
    }
};

#endif // MB_MASTER_CHAIN_H

File: built-in plugins dsp/mb-master-chain.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMasterChainPlugin: PluginDefinition = {
  id: "mb-master-chain",
  slug: "mb-master-chain",
  name: "MB Master Chain",
  category: "effect",
  type: "mastering" as any,
  version: "1.0.0",
  description: "All-in-one mastering chain with EQ, compression, and limiting",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "eqLow",
      name: "EQ Low",
      type: "float",
      defaultValue: 0,
      minValue: -6,
      maxValue: 6,
      automatable: true,
    },
    {
      id: "eqHigh",
      name: "EQ High",
      type: "float",
      defaultValue: 0,
      minValue: -6,
      maxValue: 6,
      automatable: true,
    },
    {
      id: "compThreshold",
      name: "Comp Threshold",
      type: "float",
      defaultValue: -12,
      minValue: -30,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "limCeiling",
      name: "Limiter Ceiling",
      type: "float",
      defaultValue: -0.3,
      minValue: -3,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "output",
      name: "Output",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
  ],
  defaultPreset: {
    eqLow: 0,
    eqHigh: 0,
    compThreshold: -12,
    limCeiling: -0.3,
    output: 0,
  },
};

export default MbMasterChainPlugin;
